search/query/conjunction.go

```diff
@@ -92,8 +102,9 @@ func (q *ConjunctionQuery) Validate() error {
 
 func (q *ConjunctionQuery) UnmarshalJSON(data []byte) error {
 	tmp := struct {
-		Conjuncts []json.RawMessage `json:"conjuncts"`
-		Boost     *Boost            `json:"boost,omitempty"`
+		Conjuncts  []json.RawMessage `json:"conjuncts"`
+		Boost      *Boost            `json:"boost,omitempty"`
+		ArrayDepth int               `json:"array_depth,omitempty"`
```

Comment:
One of the thoughts... is that array-depth feels like a number that goes "downwards"...

Like, given a candidate document, are the candidate terms within some allowed depth, as measured from the root or top of the document?  (That is, if I understand arrayDepth correctly -- which I might not!)

I was also playing around with thoughts of going the opposite direction, or using a relative param.

Like, given a candidate document, and you see a matching term -- how many other terms are nearby "relatives" -- kind of pretending that the document hierarchy represents a family tree.

For example, perhaps I only want to find immediate siblings (orderline product == "t-shirt" AND same orderline's size is "x-large").  But, please ignore those first cousins and second cousins.

Or, maybe siblings & first cousins are (find "Bill" and "Clinton" within the same chapter), but 2nd cousins are not good enough ("Bill" and "Clinton" mentioned by books on the same shelf isn't good enough).

Something like that.  That is, wondering if that conceptual direction would handle more generic cases (trying to hit the == and <= intuitions mentioned earlier by @sreekanth-cb and @mschoch before).

And, of course, if that works, g'luck naming such a thing!  :-)

Reply:
@steveyen , on further thinking - i feel some impedance mismatch here with this new hierarchical relation among the matched terms in a document Vs the current PR changes.

1. Though it looked alike on first read, Its slightly different from the original issue we intend to address through this PR. Here we are specifically focused on/processing term matches using their array positions as a yard stick. Or the approach won't work if we have an empty array position lists. But even in that case, your proposed hierarchical match criteria is valid and feel that is a more generic or broader use case of this approach. Current approach now feels very narrow or tailor made for a particular problem at hand. 
Since we don't have any other hierarchical context information than array position lists,  not sure how to approach this in such a generic way as don't have the full document with us. (also for documents matches where array positions aren't applicable).

2. When we think of figuring sibling relations from a bottom up manner - it feels like that mandates more/exact know how from the user on the whole document structure so that he can specify a level (int) information from the deepest level of his interest to the highest level.?
Where in the current approach, leaves more leeway for user to just convey things from a top down way. For eg: at least this deep and at more depth (could be unknown doc structure ).

3. Though initially, I had qualms over the arrayDepth name, now upon re visiting this PR after a few days - it looks more reasonable as the whole checks/verification revolves around the array position lists. This PR/approach targets only the documents that have sub arrays within it..

Reply:
The problem with "array depth" is that it's all about how we think about the implementation, and neither the name, nor the exact behavior will really make sense to a user.

I like @steveyen's proposal from that perspective.

The big concern I have is that "relative" has to be rooted somewhere, and as we've described it so far it's a just a bag of hits we're filtering, so there is no logical starting place.  The only variation I've come up with is that we could compute like a max distance.  Like you could find the deepest and shallowest hits and consider it a hit if maxD - minD was within that range (or 0 for exact same depth).

Any thoughts on our thoughts @steveyen?

Reply:
@mschoch , what if there is no array positions available in the document match? I mean the document just had a hierarchy. But there were no [] of objects. Not clear on how do we get the sense of depth for a term's occurrence?